src/platforms/esp32/components/avm_builtins/uart_driver.c

```diff
@@ -342,6 +362,15 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
         port_send_reply(ctx, pid, ref, ok_tuple);
 
     } else {
+        if (term_is_tuple(msg)) {
+            term timeover = term_get_tuple_element(msg, 1);
+            if (UNLIKELY(!term_is_integer(timeover))) {
+                globalcontext_send_message(glb, local_pid, BADARG_ATOM);
+            }
+            uint64_t timeout = ((uint64_t) term_to_int(timeover) * 1000u) + esp_timer_get_time();
+            uart_data->timeout = timeout;
+            xTaskCreate(vTaskReadTimeout, "timeout", 1000, uart_data, 1, NULL);
```

Comment:
Since you're adding esp_timer, let's not just use it to get the time.

You could use esp_timer_create.

Alternatively, you could move the timer logic up to Erlang and change the API of the port.